server/sublist.go

```diff
@@ -262,23 +292,36 @@ func (s *Sublist) Match(subject string) *SublistResult {
 	// FIXME(dlc) - Make shared pool between sublist and client readLoop?
 	result := &SublistResult{}
 
-	s.Lock()
+	s.RLock()
 	matchLevel(s.root, tokens, result)
+	s.RUnlock()
 
 	// Add to our cache
-	s.cache[subject] = result
-	// Bound the number of entries to sublistMaxCache
-	if len(s.cache) > slCacheMax {
-		for k := range s.cache {
-			delete(s.cache, k)
-			break
-		}
+	s.cache.Store(subject, result)
```

Comment:
I think this introduces a race. For example, if a client subscribes between the unlock and the call to `s.cache.Store` it will be shadowed and won't be delivered messages for this subscription until the cache entry gets cleared.
Also for queue subscriptions a client could have removed a subscription (but still be connected) and be delivered messages while no longer processing it and would "steal" it from other valid subscriptions of the same queue.

The read lock should be held while calling `s.cache.Store` to ensure that the sublist won't be modified concurrently.
There may be similar issues with `Load`, although I don't see one at the moment holding the read lock is probably a good idea too.

Reply:
exactly why I disliked that Go added sync.Map. It "introduces" these kind of issues.

Reply:
Same class of problems may happen with atomic operations but they are still useful. Concurrency is hard to reason about...

Reply:
Let me take a closer look. Let me know if your production issue resolves with this fix though.